src/OpenFeatureSDK/OpenFeatureClient.cs

```diff
@@ -17,11 +18,38 @@ namespace OpenFeatureSDK
     public sealed class FeatureClient : IFeatureClient
     {
         private readonly ClientMetadata _metadata;
-        private readonly FeatureProvider _featureProvider;
-        private readonly List<Hook> _hooks = new List<Hook>();
+        private readonly ConcurrentStack<Hook> _hooks = new ConcurrentStack<Hook>();
         private readonly ILogger _logger;
         private EvaluationContext _evaluationContext;
 
+        /// <summary>
+        /// Get a provider and an associated typed flag resolution method.
+        /// <para>
+        /// The global provider could change between two accesses, so in order to safely get provider information we
+        /// must first alias it and then use that alias to access everything we need.
+        /// </para>
+        /// </summary>
+        /// <param name="method">
+        ///     This method should return the desired flag resolution method from the given provider reference.
+        /// </param>
+        /// <typeparam name="T">The type of the resolution method</typeparam>
+        /// <returns>A tuple containing a resolution method and the provider it came from.</returns>
+        private (Func<string, T, EvaluationContext, Task<ResolutionDetails<T>>>, FeatureProvider)
+            ExtractProvider<T>(
+                Func<FeatureProvider, Func<string, T, EvaluationContext, Task<ResolutionDetails<T>>>> method)
+        {

```

Comment:
It may be worth read-locking during this operation, and include the hooks and context here. The way this is implemented right now, evaluation context and hooks can change during the process of evaluation, though you've handled the provider.